src/components/Theme/index.tsx

```diff
@@ -0,0 +1,79 @@
+import React, { createContext, useContext, useEffect, useState } from 'react'
+import { ThemeProvider as StyledComponentsThemeProvider } from 'styled-components'
+import { DarkTheme, LightTheme } from '../../themes'
+
+export enum Theme {
+	Dark = 'Dark',
+	Light = 'Light',
+}
+
+type ThemeControl = [Theme, (theme: Theme) => void]
+
+const selectTheme = {
+	[Theme.Dark]: DarkTheme,
+	[Theme.Light]: LightTheme,
+}
+
+const defaultTheme = Theme.Dark
+
+const isValidTheme = (theme: string): theme is keyof typeof Theme =>
+	Object.keys(selectTheme).includes(theme)
+
+export const toTheme = (theme: string | null): Theme | null =>
+	theme && isValidTheme(theme) ? Theme[theme] : null
+
+const getUserPreferredTheme = (): Theme | null => {
+	if (!window.matchMedia) return null
+
+	const userPrefersDark = matchMedia('(prefers-color-scheme: dark)').matches
+
+	if (userPrefersDark) return Theme.Dark
+
+	const userPrefersLight = matchMedia('(prefers-color-scheme: light)').matches
+
+	if (userPrefersLight) return Theme.Light
+
+	return null
+}
+
+const themeStorageKey = 'theme'
+
+const getInitialTheme = (): Theme => {
+	const storedTheme = toTheme(localStorage?.getItem(themeStorageKey))
+	if (storedTheme) return storedTheme
+
+	const preferredTheme = getUserPreferredTheme()
+	if (preferredTheme) return preferredTheme
+
+	return defaultTheme
+}
+
+const saveSelectedTheme = (theme: Theme): void =>
+	localStorage.setItem(themeStorageKey, theme)
+
+const ThemeControlContext = createContext<ThemeControl>([
+	defaultTheme,
+	() => {
+		return
+	},
+])
+
+const ThemeProvider: React.FC = ({ children }) => {
+	const themeControl = useState<Theme>(getInitialTheme)
+	const [theme] = themeControl
+
+	useEffect(() => saveSelectedTheme(theme), [theme])
+
+	return (
+		<ThemeControlContext.Provider value={themeControl}>
```

Comment:
I don't really understand nesting these, why this way/in this order and what are the {children} in this context.

Reply:
Well, this is simply definition of `useContext()` hook from the React that dictates similar nesting - https://reactjs.org/docs/hooks-reference.html#usecontext

`children` is also special prop from the React, that represents all the content that this component may composite when it is used. More about this topic is described in the React official documentation - https://reactjs.org/docs/composition-vs-inheritance.html